service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CreateTableCopyJobsStep.java

```diff
@@ -0,0 +1,194 @@
+package bio.terra.workspace.service.resource.controlled.flight.clone.dataset;
+
+import bio.terra.cloudres.google.bigquery.BigQueryCow;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.job.JobMapKeys;
+import bio.terra.workspace.service.resource.controlled.ControlledBigQueryDatasetResource;
+import bio.terra.workspace.service.workspace.WorkspaceService;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.api.services.bigquery.Bigquery;
+import com.google.api.services.bigquery.model.Job;
+import com.google.api.services.bigquery.model.JobConfiguration;
+import com.google.api.services.bigquery.model.JobConfigurationTableCopy;
+import com.google.api.services.bigquery.model.Table;
+import com.google.api.services.bigquery.model.TableList;
+import com.google.api.services.bigquery.model.TableList.Tables;
+import com.google.api.services.bigquery.model.TableReference;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateTableCopyJobsStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(CreateTableCopyJobsStep.class);
+  public static final Duration COPY_JOB_TIMEOUT = Duration.ofHours(12);
+  private final CrlService crlService;
+  private final WorkspaceService workspaceService;
+  private final ControlledBigQueryDatasetResource sourceDataset;
+
+  public CreateTableCopyJobsStep(
+      CrlService crlService,
+      WorkspaceService workspaceService,
+      ControlledBigQueryDatasetResource sourceDataset) {
+    this.crlService = crlService;
+    this.workspaceService = workspaceService;
+    this.sourceDataset = sourceDataset;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
```

Comment:
If we have a pod crash in the middle of launching the jobs, on restart this step would start the same copy jobs over. IIUC, the WRITE_TRUNCATE would empty the table before the copy. Does BQ make some guarantee about only one copy job running on one table at a time?  If not, then might still get duplicate rows.

In any case, could you capture the failure-recovery design in a comment in here so the reasoning is clear to The Unknown Future Maintainer?

Reply:
As @zloery points out, the working map is never persisted, so on restart we would create a new one.

Reply:
fixed by using step rerun status after each job is started.